### DISCUSSION TITLE
Firmware 20230327 new Cooldown & Stepper Disable feature

### DISCUSSION THREAD

**Discussion by I-wish-2** [mriscoc/Ender3V2S1#838] [Q&A] [answered]
Greetings,

I have been working with the current firmware reversion, { 20230327 }, Thank you Miguel for a job well done again.  I was very pleased to see that one of the new features is the Automatic Cooldown and Motor Disable have been removed.  This feature is useful to me as I want my printer to remain temperature and position stable during the time my print is cooling and being removed from my build plate which takes about 10 minutes.  I can then start another part print without waiting for the printer to stabilize again.

I did have some different results regarding the these features as noted below.  For reference I have included my end G-code and Configurator NCDF.json file.

Current Firmware: ( 20230327 )  Configurator.pyw:  E3PRO-2-427-BLTUBL-IS-LA-MPC-NCDF

; Ender 3 Pro #2 Custom END G-code 4/1/2023

G91		                 ;Relative positioning
G1 E-2 F2700	         ;Retract 1.5 mm
G1 E-2 Z0.2 F2400	;Retract and raise Z
G1 X5 Y5 F3000	;Wipe out
G1 Z10		        ;Raise Z more
G90		                ;Absolute positioning

G1 X10 Y225 Z75	;Park X and Z axis; Move Y to Present Print
M106 S128		;Set fan to 50% for cooling the hot end down to idle state.
M104 S165		;Leave Hotend On
M140 S80		;Leave Bed On
		                ;Do not disable the steppers to stop Z axis drifting down

; End of custom END G-Code

My results at the end of the print were:

The hot end temperature was set to "0"
The bed was set to "80"
The cooling fan was set to "50%"
All steppers are enabled.

My results at the end of the DEFAULT_STEPPER_DEACTIVE_TIMER 600 second time out were:
"Configuration_adv.h  line 1127 #define DEFAULT_STEPPER_DEACTIVE_TIME 600"
 
The hot end temperature remained "0"
The bed remained "80"
The cooling fan remained "50%"
All steppers were disabled at the 10 minute time out as expected.

I might note the NCDF.json file is now listed in the " Features: " section of the Configurator, but it does not make any of the changes in the firmware when it is selected.  I needed to make the three edits manually, which is ok but having the Configurator do the edits would be ideal.  I am at a loss as to why it does not work from the Configurator.  Below is the listing for the NCDF.json file built from previous work with Miguel, thank you sir.

{
"configuration.h" : [
{
  "op": "InsertAfter",
  "searchfor": "#define DWIN_LCD_PROUI",
  "newline": "#define NOT_COOL_OR_DISABLE_AT_FINISH n"
 }
],
"configuration_adv.h" : [
{
  "op": "CustomVal",
  "searchfor": "DEFAULT_STEPPER_DEACTIVE_TIMER",
  "value": "600"
 },
 {
  "op": "Disable",
  "searchfor": "#define SD_FINISHED_STEPPERRELEASE",
  "value": "false"
 }
],
"Version.h" : [
]
}

It would be great if the hotend temperature was left at the End G-Code value.  Everything could be set to "0" at the end of the  DEFAULT_STEPPER_DEACTIVE_TIMER as a safety feature.  If anyone else has seen this result please add comments.  Thanks for reading.
Regards
Doug

**Comment by mriscoc** [comment, marked as answer]
Current NCDF.json
```json
{
"Configuration.h" : [
],
"Configuration_adv.h" : [
  {
    "op": "CustomVal",
    "searchfor": "DEFAULT_STEPPER_DEACTIVE_TIME",
    "value": "300"
  },
  {
    "op": "Disable",
    "searchfor": "SD_FINISHED_STEPPERRELEASE"
  }
],
"Version.h" : [
]   
}
```
NOT_COOL_OR_DISABLE_AT_FINISH was deprecated in the latest version because now the ProUI doesn't force you to disable the motors or cooldown. Of course, you can easily cooldown and disable motors in the end G-code script.

Marlin sends the G-code M84 at the end of the printing if SD_FINISHED_STEPPERRELEASE is enabled, disabling the stepper drivers. Also, the heaters are cooled down after the DEFAULT_STEPPER_DEACTIVE_TIME.

**Reply by I-wish-2** [reply]
Thanks for the information Miguel as I have now edited the NCDF file to reflect your suggestions.  Although it shows up in the Configurator in the Features section it does not edit the configuration_Adv.h file like the other *.json files ???

I have now edited my End G-Code to reflect all the heater and motor functions that I need.  Thanks again
D

**Reply by mriscoc** [reply]
Is there any error reported in the Configurator? the NCDF.json file will disable (add "//") to **#define SD_FINISHED_STEPPERRELEASE** and set 300 to the value of **DEFAULT_STEPPER_DEACTIVE_TIME** in the Configuration_adv.h file.

**Reply by I-wish-2** [reply]
Using the information you provide in Discussion #838 and Notepad++ I typed in the lines and named the file NCDF.json and inserted it in the _features folder.  When trying to generate the Configuration these errors were displayed.

1.  -Process MyConfiguration/Configuration.h
Expecting ' , ' delimiter : line 17 column 45 (char 337)
>>>> While processing file: _features/NCDF.json

2.  -Process MyConfiguration/Configuration.h
Expecting property name enclosed in double quotes: line 17 column 46 (char 338)
>>>> While processing file: _features/NCDF.json

3.  -Process Ender3V2-427-BLTUBL-IS-LA-MPC-NCDF/Configuration.h
Expecting ' : ' delimiter : line 18 column 2 (char 347)
>>>> While processing file: _features/NCDF.json

4.  -Process MyConfiguration/Configuration.h
Expecting value: line 19 column 1 (char 349)
>>>> While processing file: _features/NCDF.json

After addressing each error in order, I was able to arrive at this NCDF.json file.

{
"configuration.h" : [
{
  "op": "InsertAfter",
  "searchfor": "#define DWIN_LCD_PROUI",
  "newline": "#define NOT_COOL_OR_DISABLE_AT_FINISH n"
 }
],
"configuration_adv.h" : [
{
  "op": "CustomVal",
  "searchfor": "DEFAULT_STEPPER_DEACTIVE_TIMER",
  "value": "600"
 },
 {
  "op": "Disable",
  "searchfor": "#define SD_FINISHED_STEPPERRELEASE",
  "value": "false"
 }
],
"Version.h" : [
]
}

I could then assemble the configuration and generate it including the NCDF addition but the changes were not made in the configuration.h or the configuration_adv.h files.
I had to enter them manually.
D

**Reply by mriscoc** [reply]
> {
> "configuration.h" : [
> {
> "op": "InsertAfter",
> "searchfor": "#define DWIN_LCD_PROUI",
> "newline": "#define NOT_COOL_OR_DISABLE_AT_FINISH n"  `<<<< INVALID`
> }
> ],
> "configuration_adv.h" : [
> {
> "op": "CustomVal",
> "searchfor": "DEFAULT_STEPPER_DEACTIVE_TIMER",  `<<<< is DEFAULT_STEPPER_DEACTIVE_TIME`
> "value": "600"
> },
> {
> "op": "Disable",
> "searchfor": "#define SD_FINISHED_STEPPERRELEASE",
> "value": "false"  `<<<< INVALID`
> }
> ],
> "Version.h" : [
> ]
> }

Re-check your json, it is invalid. Maybe there is a copy&paste error in the json file that I proposed. I zip it and post here:
[NCDF.zip](https://github.com/mriscoc/Ender3V2S1/files/11154535/NCDF.zip)
I tested the file and it is ok, if you got an error, maybe your configuration files or other json are wrong.

**Reply by I-wish-2** [reply]
Greetings, I have installed your NCDF in the _features folder and the compiler generates the configuration file ok.  Thank you.  I'm testing the compiling now.
